### PR TITLE
Add support for posting messages and replies

### PR DESCRIPTION
**Pull request by uzi**
Changed the handler to use a new postMessage action
which can handle both new posts and replies to
existing posts.  Most of the rest of the patch is
presentation.

👓 @schwers @phil303

### CODE REVIEW

File: src/app/actions/mail.js [hunk 1]
```diff
@@ -23,14 +25,40 @@ export const setInboxFailure = (mailType, error) => ({
...
 
-export const fetchInbox = (mailType, threadId) => async (dispatch, getState) => {
+export const ADD_REPLY = 'MAIL__ADD_REPLY';
+export const addReply = data => ({
+  type: ADD_REPLY,
+  data,
+});
+
+export const postMessage = (apiOptions, data) => async (dispatch, getState) => {
+  const apiResponse = await endpoints.MessagesEndpoint.post(apiOptions, data);
```
Comment (uzi) [inline src/app/actions/mail.js]: Don't call `apiResponse`

File: src/app/actions/mail.js [hunk 2]
```diff
@@ -23,14 +25,40 @@ export const setInboxFailure = (mailType, error) => ({
...
+    const data = {
+      messages: {
+        [apiResponse.name]: apiResponse,
+        [apiResponse.parentId]: newParent,
+      },
+    };
+    dispatch(addReply(data));
+  } else {
+    dispatch(platformActions.navigateToUrl(METHODS.GET, '/message/messages'));
+  }
```
Comment (uzi) [inline src/app/actions/mail.js]: Currently relies on the page not being cached

File: src/app/actions/mail.js [hunk 3]
```diff
@@ -23,14 +25,40 @@ export const setInboxFailure = (mailType, error) => ({
...
 });
 
-export const fetchInbox = (mailType, threadId) => async (dispatch, getState) => {
+export const ADD_REPLY = 'MAIL__ADD_REPLY';
+export const addReply = data => ({
+  type: ADD_REPLY,
+  data,
+});
+
+export const postMessage = (apiOptions, data) => async (dispatch, getState) => {
```
Comment (uzi) [inline src/app/actions/mail.js]: A better name than  `postMessage`?

File: src/app/actions/mail.js (hunk 1 above)
Comment (uzi) [inline src/app/actions/mail.js]: Done -- now it's `message`

File: src/app/components/MessageThread/index.jsx [hunk 4]
```diff
@@ -2,51 +2,139 @@ import './styles.less';
...
+
+  render () {
+    const {
+      messages,
+      threadId,
+    } = this.props;
+    const { showForm } = this.state;
+
+    const headId = `t4_${threadId}`;
+    const head = messages[headId];
```
Comment (uzi) [inline src/app/components/MessageThread/index.jsx]: Done

**Comment by uzi** [comment]
💇

File: src/app/reducers/mail.js [hunk 5]
```diff
@@ -12,6 +12,7 @@ const DEFAULT = [
   [mailType]: {
     pending: false,
     order: [],
+    meta: {},
```
Comment (phil303) [inline src/app/reducers/mail.js]: can you make sure to update the tests to reflect these changes?

File: src/app/reducers/messages.js [hunk 6]
```diff
@@ -15,6 +15,12 @@ export default function messages(state=DEFAULT, action={}) {
       const { messages } = action.apiResponse;
       return mergeAPIModels(state, messages);
     }
+
+    case mailActions.ADD_REPLY: {
```
Comment (phil303) [inline src/app/reducers/messages.js]: here too

File: src/app/components/MessageThread/index.jsx [hunk 7]
```diff
@@ -55,10 +138,14 @@ const selector = createSelector(
   })
 );
 
+const mapDispatchToProps = dispatch => ({
+  _onFormSubmit: data => dispatch(mailActions.postMessage(data)),
```
Comment (phil303) [inline src/app/components/MessageThread/index.jsx]: why is this prefixed with an underscore?

**Comment by phil303** [comment]
:nail_care:

File: src/app/components/MessageThread/Message.jsx [hunk 8]
```diff
@@ -27,7 +28,7 @@ export default function MessageThreadMessage(props) {
       <RedditLinkHijacker>
         <div
           className='MessageThreadMessage__body'
-          dangerouslySetInnerHTML={ { __html: message.bodyHTML } }
+          dangerouslySetInnerHTML={ { __html: mobilify(message.bodyHTML) } }
```
Comment (schwers) [inline src/app/components/MessageThread/Message.jsx]: 🔕  Just as an aside, I wonder if we should even be using `mobilify` anymore. It really only does url conversion as @phil303 said, but I think we should make `RedditLinkHijacker` robust enough that devs don't have to worry about using `mobilify`. The only reason I can think of _needing_ to use `mobilify` is if it converted urls to canonical urls (at least in terms of https, domain, etc) for SEO reasons.

File: src/app/router/handlers/DirectMessage.js [hunk 9]
```diff
@@ -1,20 +1,16 @@
...
+import * as platformActions from '@r/platform/actions';
 
 import { fetchUserBasedData } from './handlerCommon';
-import { apiOptionsFromState } from 'lib/apiOptionsFromState';
 import { getBasePayload, logClientScreenView } from 'lib/eventUtils';
 
 export default class DirectMessage extends BaseHandler {
   async [METHODS.GET](dispatch, getState) {
+    if (!getState().session.isValid) {
+      return dispatch(platformActions.setPage('/register'));
```
Comment (phil303) [inline src/app/router/handlers/DirectMessage.js]: just a heads up, you can now leverage the [constant](https://github.com/reddit/reddit-mobile/blob/2X/src/app/constants.js#L91) here

File: src/app/components/MessageThread/index.jsx [hunk 10]
```diff
@@ -2,51 +2,134 @@ import './styles.less';
...
+  onToggleShowForm = () => {
+    this.setState({ showForm: !this.state.showForm });
+  };
+
+  onEditReply = (e) => {
+    e.preventDefault();
+    this.setState({ formBody: e.target.value });
+  };
+
+  render () {
```
Comment (schwers) [inline src/app/components/MessageThread/index.jsx]: Nit: `render`, `renderMessage`, `renderReplyToggle`, and `renderForm` shouldn't have a space before parens

**Comment by schwers** [comment]
💅

**Comment by uzi** [comment]
💇 Everything should be taken care of now with the exception of mail reducer tests... there aren't any.  I'm putting this out so you guys can review the rest of it and I'll go do that one remaining thing now.

**Comment by phil303** [comment]
@uzi can you make sure to address [this comment](https://github.com/reddit/reddit-mobile/pull/805/files#r82487305) as well

**Comment by uzi** [comment]
💇 Tests added... should be good to go.

**Comment by schwers** [comment]
:fish:

**Comment by phil303** [comment]
you don't need it from me but :fish:!

**Pull request merged**